### PR TITLE
NIFI-1123 Adds expression language support to DeleteAttributesExpress… … …ion on UpdateAttributes Processor.

### PR DESCRIPTION
**Pull request by jskora**
NIFI-1123 This branch contains changes to support Expression Language in the UpdateAttributes processor DeleteAttributesExpression property (originally discussed on [NIFI-641](https://issues.apache.org/jira/browse/NIFI-641)) and related unit tests.

### CODE REVIEW

File: nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java [hunk 1]
```diff
@@ -136,7 +136,8 @@
             .name("Delete Attributes Expression")
             .description("Regular expression for attributes to be deleted from flowfiles.")
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java]: Does the description need to change here?

**Comment by trkurc** [comment]
@jskora what would you expect the following unit tests to do?

```
    @Test
    public void testInvalidRegex() {
        final TestRunner runner = TestRunners.newTestRunner(new UpdateAttribute());
        runner.setProperty("Delete Attributes Expression", "(");

        final Map<String, String> attributes = new HashMap<>();
        attributes.put("attribute.1", "value.1");

        runner.enqueue(new byte[0], attributes);

        runner.run();

    }
    
    @Test
    public void testInvalidRegexInAttribute() {
        final TestRunner runner = TestRunners.newTestRunner(new UpdateAttribute());
        runner.setProperty("Delete Attributes Expression", "${butter}");

        final Map<String, String> attributes = new HashMap<>();
        attributes.put("butter", "(");

        runner.enqueue(new byte[0], attributes);

        runner.run();

    }

```

**Comment by trkurc** [comment]
I think what I'm getting at is that I think having a property that requires the expression to evaluate to a valid regex might mean it is time for a failure relationship for this processor (a breaking change?)

**Comment by apiri** [comment]
@trkurc I am onboard with what you are driving toward. I had created an issue around the same thought process. https://issues.apache.org/jira/browse/NIFI-813

The core issue is that EL, when introduced, provided ways for processors to fail in ways previously unanticipated.

**Comment by jskora** [comment]
Interesting.  Before I enabled EL support, it had a REGULAR_EXPRESSION_VALIDATOR which would have caught those.  Is there a way to validate a property that supports expression language to verify that the output will be an attribute key?

**Comment by trkurc** [comment]
@jskora as @apiri pointed out, i think that we may have to live with the problem until NIFI-813 is resolved. I tried wrapping my head around a validator that handled the first test case, and that hurt. The second one didn't hurt, because a validator is too early to catch.

**Comment by trkurc** [comment]
@apiri you have some history with this - what is a greater evil, no el support or possibly breaking? this is already a processor that appears to have a "beware of el" sign up already?

**Comment by jskora** [comment]
@trkurc, this is not a problem fix but an enhancement based on discussion that occurred on [NIFI-641|https://issues.apache.org/jira/browse/NIFI-641].  It seemed easy enough, but the validation issues probably out weigh the value.

Postponing until [NIFI-813|https://issues.apache.org/jira/browse/NIFI-813] provides a means to handle potential errors, but if mixing Regex and Expression Language can only be handled with a Failure relationships, maybe it's better to not mix the them and allow separate properties so the Regex can be validated and the EL expression can produce a list of attributes that can be ignored if they don't exist.   Thoughts?

 until they can be added without adding risk.

**Comment by trkurc** [comment]
@jskora - I think my question to @apiri  was "is this an excusable problem". If I read that other ticket right, other configurations already exhibit similar behavior. Multiple properties might be a greater evil.

**Comment by apiri** [comment]
@trkurc Wasn't trying to divert away from the feature add; totally supportive of that.  Was just trying to link Tony's thought process with something I stumbled across before.

Personally, I think it is something we can/have to let slide for now, and then figure out a reasonable path forward of incorporating these generated relationships across the processor set as we go to 1.0.  Long term, when someone opts for an EL perhaps the framework can provide a new relationship dynamically or something similar.

**Comment by jskora** [comment]
I like that a bad regex stops a processor from being started, so I'd hate to lose that.

**Comment by apiri** [comment]
Was combing through the actual PR.  Not sure I am onboard with the additional property and misunderstood what the intent was.  There was a similar effort with the Character Set Validator where we validated that the specified charset was valid if specified, otherwise just that a string was returned via EL should it be an expression.

Regex Validator, by default evaluates the EL and then applies validation to the resulting "regex".  Would it be reasonable to have a separate validator that matches the semantics of what you are looking for (akin to that of how ConvertCharacterSet functions)?  Should be okay from the standpoint of compatibility and non-breaking changes.

Additionally, playing with this a bit, I did a blanket delete (.*) which leads to some bad effects in terms of the framework as the UUID is removed and puts us in a bad state.

**Comment by markap14** [comment]
Yikes! @apiri the fact that the uuid attribute can be removed is a bug. I created a ticket for this: https://issues.apache.org/jira/browse/NIFI-1182

I do agree with your statements above.

**Comment by jskora** [comment]
@apiri, the property value is compiled as a regex, any separators are in the context of the regex.  

I'm not sure I understand how the RegexValidator can evaluate the EL since it could refer to dynamic attributes?  If the validator works, won't the UI prevent the processor from being started with the bad values like those in @trkurc's tests?  And won't the tests fail to run the processor since it is invalid?  If that's the case, the invalid values will never be used.

Regarding the deletion of framework attributes, should UpdateAttributes prevent their deletion or should the ProcessSession sub-classes prevent that more globally?  What framework attributes should be protected, is it everything in CoreAttributes?

**Comment by apiri** [comment]
@jskora Sorry, I think a few of the different contexts from messages between you, @trkurc and I made communication a bit jumbled.

To that end, I am onboard with regex, but personally, would not want a second field to the same focus.  The issue that is present is more of an issue with the underlying framework and its handling with EL.  What I was suggesting with the regex validator is applying the validation to an entered string, but bypassing validation for where these values are specified on a per FlowFile basis via EL.  What is missing is not so much the granularity as it is the ability to route failures which is currently nonexistent and was some of what @trkurc was asking about what the appropriate tradeoffs/concessions are given the current state of the framework.  To this idea, I am more for keeping consistent with how we are now and letting this get addressed with the needed framework adjustments as we move forward.

To your question regarding deletion, that is certainly not on you or any extension developer.  As highlighted by NIFI-1181/1182, these values are safeguarded but your processor exposed the bug of where it was not safeguarded in this context.

**Comment by trkurc** [comment]
I am 100% in agreement with @apiri

**Comment by jskora** [comment]
@apiri, that makes sense regarding the framework.  But, in the meantime, it would be an easy patch to add an exclusion list to UpdateAttributes to prevent the deletes as a stop gap measure, if that would help.

**Comment by apiri** [comment]
Should be a quick fix and should go in for 0.4.0 so would opt to just keep your processor free of that logic.

**Comment by trkurc** [comment]
@apiri if you're +1, I can merge this in.

**Comment by apiri** [comment]
@trkurc My only lingering point is that of the custom validator for this field.  Validate non-EL input as a regular expression, when EL just validate the expression results in a returned string.

Thoughts?

**Comment by jskora** [comment]
@apiri and @trkurc, I can add the custom validator and submit a new PR.  Is there a standard way to determine whether a property includes EL?

**Comment by apiri** [comment]
You could do something similar to https://github.com/apache/nifi/blob/master/nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/StandardValidators.java#L208

**Comment by jskora** [comment]
Ok, I think that commit fixes the validation issue.  This commit
1. adds a custom validator checking that the expression is either valid EL or valid RE, and
2. adds tests to match Tony's bad RE tests.

**Comment by apiri** [comment]
Awesome and thanks! Will put this on the queue to review if @trkurc doesn't beat me to it.

n.b. The adjustments for guarding the UUID attribute were merged in earlier today.

**Comment by trkurc** [comment]
@jskora - this was merged in  but I neglected to add the "This closes #116" to the commit message.

**Comment by jskora** [comment]
Do you need me to close the pull request?

**Comment by trkurc** [comment]
@jskora - yes, please!

**Comment by jskora** [comment]
Closed by apache/nifi commits 52b24b93d9f7763744c792c0cfed8974f8e6cb83 and 9e2f6df20511b814b761726e40f2d3b1f498cc9f.

**Pull request closed without merging**